ta/pkcs11/src/pkcs11_token.c

```diff
@@ -111,3 +111,54 @@ uint32_t entry_ck_slot_list(uint32_t ptypes, TEE_Param *params)
 
 	return PKCS11_CKR_OK;
 }
+
+uint32_t entry_ck_slot_info(uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = &params[0];
+	TEE_Param *out = &params[2];
+	uint32_t rv = 0;
+	struct serialargs ctrlargs = { };
+	uint32_t token_id = 0;
+	struct ck_token *token = NULL;
+	const char desc[] = PKCS11_SLOT_DESCRIPTION;
+	const char manuf[] = PKCS11_SLOT_MANUFACTURER;
+	const char hwver[2] = PKCS11_SLOT_HW_VERSION;
+	const char fwver[2] = PKCS11_SLOT_FW_VERSION;
+	struct pkcs11_slot_info info = { };
+
+	if (ptypes != exp_pt ||
+	    out->memref.size != sizeof(struct pkcs11_slot_info))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rv = serialargs_get(&ctrlargs, &token_id, sizeof(uint32_t));
+	if (rv)
+		return rv;
+
+	if (serialargs_remaining_bytes(&ctrlargs))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	token = get_token(token_id);
+	if (!token)
+		return PKCS11_CKR_SLOT_ID_INVALID;
+
+	PADDED_STRING_COPY(info.slot_description, desc);
```

Comment:
This is already zero initialized, why zero initialize the unused bytes again?

Reply:
The spec expects padding with space chars, not zeroes hence a specific padding helper.

Reply:
I noticed that we're padding in both this struct and in another struct in libckteec, do we really need to do it twice?

Reply:
libckteec pads only local strings. The string identifiers output by the TA are not re-padded by libckteec.